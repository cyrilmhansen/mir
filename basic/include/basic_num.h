#ifndef BASIC_NUM_H
#define BASIC_NUM_H

#include <math.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef enum {
  BASIC_NUM_MODE_DOUBLE,
  BASIC_NUM_MODE_LONG_DOUBLE,
  BASIC_NUM_MODE_FIXED64,
} basic_num_mode_t;

static inline void basic_num_init (basic_num_mode_t mode) { (void) mode; }

#if defined(BASIC_USE_LONG_DOUBLE)

typedef long double basic_num_t;
#define BASIC_ZERO ((basic_num_t) 0.0L)
#define BASIC_FROM_INT(x) ((basic_num_t) (x))
#define BASIC_TO_INT(x) ((long) (x))
#define BASIC_STRTOF strtold
#define BASIC_NUM_SCANF "%Lf"
#define BASIC_NUM_PRINTF "%.18Lg"
#define BASIC_ADD(a, b) ((a) + (b))
#define BASIC_SUB(a, b) ((a) - (b))
#define BASIC_MUL(a, b) ((a) * (b))
#define BASIC_DIV(a, b) ((a) / (b))
#define BASIC_NEG(a) (-(a))
#define BASIC_EQ(a, b) ((a) == (b))
#define BASIC_NE(a, b) ((a) != (b))
#define BASIC_LT(a, b) ((a) < (b))
#define BASIC_LE(a, b) ((a) <= (b))
#define BASIC_GT(a, b) ((a) > (b))
#define BASIC_GE(a, b) ((a) >= (b))
#define BASIC_FABS fabsl
#define BASIC_SQRT sqrtl
#define BASIC_SIN sinl
#define BASIC_COS cosl
#define BASIC_TAN tanl
#define BASIC_SINH sinhl
#define BASIC_COSH coshl
#define BASIC_TANH tanhl
#define BASIC_ASINH asinhl
#define BASIC_ACOSH acoshl
#define BASIC_ATANH atanhl
#define BASIC_ASIN asinl
#define BASIC_ACOS acosl
#define BASIC_ATAN atanl
#define BASIC_LOG logl
#define BASIC_LOG2 log2l
#define BASIC_LOG10 log10l
#define BASIC_EXP expl
#define BASIC_POW powl
#define BASIC_FLOOR floorl

#elif defined(BASIC_USE_FIXED64)

#include <ctype.h>
#include <errno.h>
#include <stdint.h>
#include "fixed64/fixed64.h"

typedef fixed64_t basic_num_t;
#define BASIC_ZERO ((basic_num_t) {0, 0})
#define BASIC_FROM_INT(x) fixed64_from_int (x)
#define BASIC_TO_INT(x) fixed64_to_int (x)
#define BASIC_STRTOF fixed64_from_string
#define BASIC_NUM_SCANF "%s"
#define BASIC_NUM_PRINTF "%s"
#define BASIC_ADD(a, b) fixed64_add ((a), (b))
#define BASIC_SUB(a, b) fixed64_sub ((a), (b))
#define BASIC_MUL(a, b) fixed64_mul ((a), (b))
#define BASIC_DIV(a, b) fixed64_div ((a), (b))
#define BASIC_NEG(a) fixed64_neg (a)
#define BASIC_EQ(a, b) (fixed64_cmp ((a), (b)) == 0)
#define BASIC_NE(a, b) (fixed64_cmp ((a), (b)) != 0)
#define BASIC_LT(a, b) (fixed64_cmp ((a), (b)) < 0)
#define BASIC_LE(a, b) (fixed64_cmp ((a), (b)) <= 0)
#define BASIC_GT(a, b) (fixed64_cmp ((a), (b)) > 0)
#define BASIC_GE(a, b) (fixed64_cmp ((a), (b)) >= 0)
#define BASIC_FABS fixed64_abs
#define BASIC_SQRT fixed64_sqrt
#define BASIC_SIN fixed64_sin
#define BASIC_COS fixed64_cos
#define BASIC_TAN fixed64_tan
#define BASIC_SINH fixed64_sinh
#define BASIC_COSH fixed64_cosh
#define BASIC_TANH fixed64_tanh
#define BASIC_ASINH fixed64_asinh
#define BASIC_ACOSH fixed64_acosh
#define BASIC_ATANH fixed64_atanh
#define BASIC_ASIN fixed64_asin
#define BASIC_ACOS fixed64_acos
#define BASIC_ATAN fixed64_atan
#define BASIC_LOG fixed64_log
#define BASIC_LOG2 fixed64_log2
#define BASIC_LOG10 fixed64_log10
#define BASIC_EXP fixed64_exp
#define BASIC_POW fixed64_pow
#define BASIC_FLOOR fixed64_floor

#else /* BASIC_USE_DOUBLE */

typedef double basic_num_t;
#define BASIC_ZERO ((basic_num_t) 0.0)
#define BASIC_FROM_INT(x) ((basic_num_t) (x))
#define BASIC_TO_INT(x) ((long) (x))
#define BASIC_STRTOF strtod
#define BASIC_NUM_SCANF "%lf"
#define BASIC_NUM_PRINTF "%.15g"
#define BASIC_ADD(a, b) ((a) + (b))
#define BASIC_SUB(a, b) ((a) - (b))
#define BASIC_MUL(a, b) ((a) * (b))
#define BASIC_DIV(a, b) ((a) / (b))
#define BASIC_NEG(a) (-(a))
#define BASIC_EQ(a, b) ((a) == (b))
#define BASIC_NE(a, b) ((a) != (b))
#define BASIC_LT(a, b) ((a) < (b))
#define BASIC_LE(a, b) ((a) <= (b))
#define BASIC_GT(a, b) ((a) > (b))
#define BASIC_GE(a, b) ((a) >= (b))
#define BASIC_FABS fabs
#define BASIC_SQRT sqrt
#define BASIC_SIN sin
#define BASIC_COS cos
#define BASIC_TAN tan
#define BASIC_SINH sinh
#define BASIC_COSH cosh
#define BASIC_TANH tanh
#define BASIC_ASINH asinh
#define BASIC_ACOSH acosh
#define BASIC_ATANH atanh
#define BASIC_ASIN asin
#define BASIC_ACOS acos
#define BASIC_ATAN atan
#define BASIC_LOG log
#define BASIC_LOG2 log2
#define BASIC_LOG10 log10
#define BASIC_EXP exp
#define BASIC_POW pow
#define BASIC_FLOOR floor

#endif

static inline basic_num_t basic_num_from_int (long x) { return BASIC_FROM_INT (x); }
static inline basic_num_t basic_num_from_string (const char *s, char **end) {
  return BASIC_STRTOF (s, end);
}
static inline long basic_num_to_int (basic_num_t x) { return BASIC_TO_INT (x); }
static inline basic_num_t basic_num_add (basic_num_t a, basic_num_t b) { return BASIC_ADD (a, b); }
static inline basic_num_t basic_num_sub (basic_num_t a, basic_num_t b) { return BASIC_SUB (a, b); }
static inline basic_num_t basic_num_mul (basic_num_t a, basic_num_t b) { return BASIC_MUL (a, b); }
static inline basic_num_t basic_num_div (basic_num_t a, basic_num_t b) { return BASIC_DIV (a, b); }
static inline basic_num_t basic_num_neg (basic_num_t a) { return BASIC_NEG (a); }
static inline int basic_num_eq (basic_num_t a, basic_num_t b) { return BASIC_EQ (a, b); }
static inline int basic_num_ne (basic_num_t a, basic_num_t b) { return BASIC_NE (a, b); }
static inline int basic_num_lt (basic_num_t a, basic_num_t b) { return BASIC_LT (a, b); }
static inline int basic_num_le (basic_num_t a, basic_num_t b) { return BASIC_LE (a, b); }
static inline int basic_num_gt (basic_num_t a, basic_num_t b) { return BASIC_GT (a, b); }
static inline int basic_num_ge (basic_num_t a, basic_num_t b) { return BASIC_GE (a, b); }
static inline basic_num_t basic_num_fabs (basic_num_t x) { return BASIC_FABS (x); }
static inline basic_num_t basic_num_sqrt (basic_num_t x) { return BASIC_SQRT (x); }
static inline basic_num_t basic_num_sin (basic_num_t x) { return BASIC_SIN (x); }
static inline basic_num_t basic_num_cos (basic_num_t x) { return BASIC_COS (x); }
static inline basic_num_t basic_num_tan (basic_num_t x) { return BASIC_TAN (x); }
static inline basic_num_t basic_num_sinh (basic_num_t x) { return BASIC_SINH (x); }
static inline basic_num_t basic_num_cosh (basic_num_t x) { return BASIC_COSH (x); }
static inline basic_num_t basic_num_tanh (basic_num_t x) { return BASIC_TANH (x); }
static inline basic_num_t basic_num_asinh (basic_num_t x) { return BASIC_ASINH (x); }
static inline basic_num_t basic_num_acosh (basic_num_t x) { return BASIC_ACOSH (x); }
static inline basic_num_t basic_num_atanh (basic_num_t x) { return BASIC_ATANH (x); }
static inline basic_num_t basic_num_asin (basic_num_t x) { return BASIC_ASIN (x); }
static inline basic_num_t basic_num_acos (basic_num_t x) { return BASIC_ACOS (x); }
static inline basic_num_t basic_num_atan (basic_num_t x) { return BASIC_ATAN (x); }
static inline basic_num_t basic_num_log (basic_num_t x) { return BASIC_LOG (x); }
static inline basic_num_t basic_num_log2 (basic_num_t x) { return BASIC_LOG2 (x); }
static inline basic_num_t basic_num_log10 (basic_num_t x) { return BASIC_LOG10 (x); }
static inline basic_num_t basic_num_exp (basic_num_t x) { return BASIC_EXP (x); }
static inline basic_num_t basic_num_pow (basic_num_t x, basic_num_t y) { return BASIC_POW (x, y); }
static inline basic_num_t basic_num_floor (basic_num_t x) { return BASIC_FLOOR (x); }

#if defined(BASIC_USE_FIXED64)
static inline int basic_num_scan (FILE *f, basic_num_t *out) {
  char buf[128], *end;
  if (fgets (buf, sizeof (buf), f) == NULL) return 0;
  errno = 0;
  double d = strtod (buf, &end);
  if (end == buf || errno == ERANGE) return 0;
  while (isspace ((unsigned char) *end)) end++;
  if (*end != '\0') return 0;
  if (d < (double) INT64_MIN || d > (double) INT64_MAX) return 0;
  *out = fixed64_from_double (d);
  return 1;
}
static inline void basic_num_print (FILE *f, basic_num_t x) {
  char buf[128];
  fixed64_to_string (x, buf, sizeof (buf));
  fputs (buf, f);
}
static inline int basic_num_to_chars (basic_num_t x, char *buf, size_t size) {
  return fixed64_to_string (x, buf, size);
}
#else
static inline int basic_num_scan (FILE *f, basic_num_t *out) {
  return fscanf (f, BASIC_NUM_SCANF, out) == 1;
}
static inline void basic_num_print (FILE *f, basic_num_t x) { fprintf (f, BASIC_NUM_PRINTF, x); }
static inline int basic_num_to_chars (basic_num_t x, char *buf, size_t size) {
  return snprintf (buf, size, BASIC_NUM_PRINTF, x);
}
#endif

#endif /* BASIC_NUM_H */
