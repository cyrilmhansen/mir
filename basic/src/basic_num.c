#include "basic_num.h"

static void basic_num_init_double (void);
static void basic_num_init_long_double (void);
static void basic_num_init_fixed64 (void);
static void basic_num_init_mb5 (void);

basic_num_ops_t basic_num;

void basic_num_init (basic_num_mode_t mode) {
  switch (mode) {
  case BASIC_NUM_MODE_LONG_DOUBLE: basic_num_init_long_double (); break;
  case BASIC_NUM_MODE_FIXED64: basic_num_init_fixed64 (); break;
  case BASIC_NUM_MODE_MBASIC5: basic_num_init_mb5 (); break;
  case BASIC_NUM_MODE_DOUBLE:
  default: basic_num_init_double (); break;
  }
}
