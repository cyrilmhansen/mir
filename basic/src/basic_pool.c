#include "basic_pool.h"
#include "basic_num.h"

#include <stdlib.h>
#include <string.h>
#include <stddef.h>

typedef struct PoolBlock {
  char *data;
  size_t size;
  size_t pos;
  struct PoolBlock *next;
} PoolBlock;

typedef struct FreeBlock {
  size_t size;
  struct FreeBlock *next;
} FreeBlock;

static PoolBlock *pool = NULL;
static FreeBlock *free_list = NULL;

static size_t align_up (size_t n) {
  size_t align1 = _Alignof (max_align_t);
  size_t align2 = _Alignof (basic_num_t);
  size_t align = align1 > align2 ? align1 : align2;
  return (n + align - 1) & ~(align - 1);
}

void basic_pool_init (size_t size) {
  if (pool != NULL) basic_pool_destroy ();
  free_list = NULL;
  if (size != 0) {
    pool = malloc (sizeof (PoolBlock));
    if (pool != NULL) {
      pool->data = malloc (size);
      if (pool->data == NULL) {
        free (pool);
        pool = NULL;
      } else {
        pool->size = size;
        pool->pos = 0;
        pool->next = NULL;
      }
    }
  } else {
    pool = NULL;
  }
}

static PoolBlock *curr_block (void) {
  PoolBlock *b = pool;
  if (b == NULL) return NULL;
  while (b->next != NULL) b = b->next;
  return b;
}

void *basic_pool_alloc (size_t size) {
  size = align_up (size);
  /* First, try to reuse a freed block. */
  for (FreeBlock **pb = &free_list; *pb != NULL; pb = &(*pb)->next) {
    if ((*pb)->size >= size) {
      FreeBlock *res = *pb;
      *pb = res->next;
      return (char *) res + sizeof (FreeBlock);
    }
  }
  if (pool == NULL) basic_pool_init (1024);
  PoolBlock *b = curr_block ();
  if (b == NULL) return NULL;
  size_t total = sizeof (FreeBlock) + size;
  if (b->pos + total > b->size) {
    size_t newsize = b->size ? b->size * 2 : 1024;
    while (newsize < total) newsize *= 2;
    PoolBlock *nb = malloc (sizeof (PoolBlock));
    if (nb == NULL) return NULL;
    nb->data = malloc (newsize);
    if (nb->data == NULL) {
      free (nb);
      return NULL;
    }
    nb->size = newsize;
    nb->pos = 0;
    nb->next = NULL;
    b->next = nb;
    b = nb;
  }
  FreeBlock *res = (FreeBlock *) (b->data + b->pos);
  res->size = size;
  res->next = NULL;
  b->pos += total;
  return (char *) res + sizeof (FreeBlock);
}

void basic_pool_reset (void) {
  for (PoolBlock *b = pool; b != NULL; b = b->next) b->pos = 0;
  free_list = NULL;
}

void basic_pool_destroy (void) {
  PoolBlock *b = pool;
  while (b != NULL) {
    PoolBlock *next = b->next;
    free (b->data);
    free (b);
    b = next;
  }
  pool = NULL;
  free_list = NULL;
}

char *basic_alloc_string (size_t len) {
  char *s = basic_pool_alloc (len + 1);
  if (s != NULL) s[len] = '\0';
  return s;
}

void *basic_alloc_array (size_t count, size_t elem_size, int clear) {
  size_t n = count * elem_size;
  void *p = basic_pool_alloc (n);
  if (p != NULL && clear) memset (p, 0, n);
  return p;
}

void *basic_calloc (size_t count, size_t elem_size) {
  return basic_alloc_array (count, elem_size, 1);
}

void basic_pool_free (void *p) {
  if (p == NULL) return;
  FreeBlock *b = (FreeBlock *) ((char *) p - sizeof (FreeBlock));
  b->next = free_list;
  free_list = b;
}

int basic_clear_array_pool (void *base, size_t count, size_t elem_size) {
  if (base == NULL) return 0;
  size_t n = count * elem_size;
  for (PoolBlock *b = pool; b != NULL; b = b->next) {
    if ((char *) base >= b->data && (char *) base + n <= b->data + b->size) {
      memset (base, 0, n);
      return 1;
    }
  }
  return 0;
}
