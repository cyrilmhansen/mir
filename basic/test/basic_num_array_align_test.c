#include "basic_pool.h"
#include "basic_num.h"
#include <stdint.h>
#include <stdio.h>

int main (void) {
  basic_pool_init (0);
  basic_num_init (BASIC_NUM_MODE_FIXED64);
  basic_num_t *arr = basic_calloc (4, sizeof (basic_num_t));
  if (arr == NULL) return 1;
  if ((uintptr_t) arr % _Alignof (basic_num_t) != 0) return 1;
  basic_pool_free (arr);
  basic_pool_destroy ();
  printf ("basic_num_array_align_test OK\n");
  return 0;
}
