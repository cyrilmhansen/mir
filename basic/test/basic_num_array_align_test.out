basic_num_array_align_test OK
